Added a few functions

LiamGoodacre opened the pull request:
* replicate
* replicateA
* empty
* singleton
* singletonA
* range

With some basic tests.

garyb commented:
Neat! :+1: 

I do wonder a little about `empty`, as we also have `empty` from `Control.Plus`, but I guess not every `Unfoldable` is necessarily a `Plus`, even though `Unfoldable` is the right kind for it.

On that note (unrelated to this PR) should we have a `Functor` constraint on `Unfoldable`? (@paf31?)

paf31 commented on src/Data/Unfoldable.purs:
```diff
@@ -38,3 +39,58 @@ instance unfoldableArray :: Unfoldable Array where
+-- | range 3 3 == [3] :: Array Int
+-- | range 4 2 == [] :: Array Int
+-- | ~~~
+range :: forall f. (Unfoldable f) => Int -> Int -> f Int
```
Perhaps this can be generalized from `Int` to `Enum`?

garyb replied:
I guess we need to finish up/merge the `Enum` changes for that first.

paf31 replied:
True. So many outstanding PRs to merge ..

LiamGoodacre replied:
Okay so I've just seen that pull req on purescript-enums.
Looks like purescript-enums actually depends on purescript-unfoldable too.
Should I remove `range` here and let `enumFromTo` be implemented in purescript-enums?

garyb replied:
I'm not sure which way is best, but as an aside I much prefer the name `range` :smile:

paf31 commented on src/Data/Unfoldable.purs:
```diff
@@ -38,3 +39,58 @@ instance unfoldableArray :: Unfoldable Array where
+-- | Perform an Applicative action and contain the single result.
+singletonA :: forall m f a. (Applicative m, Unfoldable f, Traversable f) =>
+              m a -> m (f a)
+singletonA = replicateA 1
```
Feels like this should be `map singleton` although we probably lack the laws to prove it. If it is, maybe it's simple enough to not warrant its own function here, especially given the relatively complex instance context.

LiamGoodacre replied:
I didn't realise that when I wrote it :P - I'll remove it.

paf31 commented on src/Data/Unfoldable.purs:
```diff
@@ -38,3 +39,58 @@ instance unfoldableArray :: Unfoldable Array where
+-- | ~~~ purescript
+-- | empty == [] :: forall a. Array a
+-- | ~~~
+empty :: forall f a. (Unfoldable f) => f a
```
I'd rather avoid having naming collisions between two core control libraries if possible, unless one is a generalization of the other. Since I'm not sure if this is a generalization of `Control.Plus.empty` in general, could we rename it, please? `nil`, or `none` perhaps?

LiamGoodacre replied:
Agreed - Renaming to `none`.

paf31 commented:
I think `Foldable` should probably have a `Functor` constraint too, no?

paf31 commented:
:+1: Looks nice, thanks!

garyb commented:
Re: `Functor <= Foldable` too - we sorta discussed it: purescript/purescript-foldable-traversable#16 but didn't actually do anything with it.

LiamGoodacre commented:
Latest update:

* renamed `empty` to `none`
* removed `singletonA`
* removed `range` - once purescript-enums has been sorted out we should revisit this

paf31 commented:
Looks good to merge to me. @garyb Any comments? Also, we could probably deprecate the equivalent methods in `arrays` and `lists` after this.

garyb commented:
:+1: and sounds good. I did some stuff with `Alternative` on `arrays`, so if we go ahead with that we can remove the equivalents then perhaps.

paf31 commented:
:+1: Thanks!

LiamGoodacre commented:
:smile: Thanks guys
